Newsletters: add 142 (2021-03-27) (#539)
bitcoinops/bitcoinops.github.io: _posts/en/newsletters/2021-03-31-newsletter.md

    | 0 sat | 100,000,000 sat |
    | 1 sat | 99,999,999 sat |
    | ... | ...|
    | 100,000,000 sat | 0 sat |

   Making this assumption allows the authors to draw conclusions about
   the probability that a payment will succeed based on its amount and how
   many hops (channels) it needs to traverse.  This allows them to
   prove the benefit of several known heuristics---such as keeping paths
   short and using [multipath payments][topic multipath payments] to break
   larger payments into smaller payments (under certain other
   assumptions).  They also use the model to evaluate new proposals,
   such as allowing [Just-In-Time (JIT) rebalancing][topic jit routing]
   via [bolts #780][].

   The paper uses its conclusions to provide a routing algorithm that it
   claims can reduce payment retry attempts by 20% compared to their
   simplification of existing routing algorithms.  The new algorithm
   prefers routes with a higher computed probability of success, whereas
   existing algorithms use a heuristic approach.  Combined with JIT
   rebalancing, they estimate a 48% improvement.  Given that each retry
   usually requires several seconds, and could take much longer in some

>    cases, this could provide a significant benefit to the network.

Reviewer: I'm not sure what "benefit to the network" means here. Would "improved user experience" be better or is it broader than that?

projects.  Please consider upgrading to new releases or helping to test
release candidates.*

- [HWI 2.0.1][] is a bugfix release that addresses minor issues with Trezor
  T passphrase entry and keyboard shortcuts in the `hwi-qt` user
  interface.

- [C-Lightning 0.10.0-rc1][C-Lightning 0.10.0] is a release candidate
  for the next major version of this LN node software.

## Notable code and documentation changes

*Notable changes this week in [Bitcoin Core][bitcoin core repo],
[C-Lightning][c-lightning repo], [Eclair][eclair repo], [LND][lnd repo],
[Rust-Lightning][rust-lightning repo], [libsecp256k1][libsecp256k1
repo], [Hardware Wallet Interface (HWI)][hwi repo],
[Rust Bitcoin][rust bitcoin repo], [BTCPay Server][btcpay server repo],
[Bitcoin Improvement Proposals (BIPs)][bips repo], and [Lightning
BOLTs][bolts repo].*

- [Bitcoin Core #17227][] Qt: Add Android packaging support FIXME:jnewbery

- [Rust-Lightning #849][] makes a channel's `cltv_expiry_delta`
  configurable and reduces the default value from 72 blocks to 36
  blocks.  This parameter sets the minimum number of blocks a node will
  have to settle a payment attempt with its upstream peer after learning

>   from its downstream peer whether or not that payment succeeded; it must

Reviewer: "This parameter [...] payment succeeded;" is a bit long and difficult to parse. "the minimum number of blocks a node will have to settle" in particular reads like the node has to settle blocks, whereas it really means "the number of blocks within which a node must settle".

   prove the benefit of several known heuristics---such as keeping paths
   short and using [multipath payments][topic multipath payments] to break
   larger payments into smaller payments (under certain other
   assumptions).  They also use the model to evaluate new proposals,
   such as allowing [Just-In-Time (JIT) rebalancing][topic jit routing]
   via [bolts #780][].

   The paper uses its conclusions to provide a routing algorithm that it
   claims can reduce payment retry attempts by 20% compared to their
   simplification of existing routing algorithms.  The new algorithm
   prefers routes with a higher computed probability of success, whereas
   existing algorithms use a heuristic approach.  Combined with JIT
   rebalancing, they estimate a 48% improvement.  Given that each retry
   usually requires several seconds, and could take much longer in some
   cases, this could provide a significant benefit to the network.
   The algorithm was tested against several example networks, including
   one drawn from a snapshot of almost 1,000 live channels.

   The paper deliberately does not take routing fees into consideration,

>    and most mailing list discussion about the topic focused on how to

Reviewer 2: perhaps: "most of the ensuing mailing list discussion focused on..."

   use the results while still ensuring users don't pay an excessive
   amount of fees.

## Selected Q&A from Bitcoin StackExchange

*[Bitcoin StackExchange][bitcoin.se] is one of the first places Optech
contributors look for answers to their questions---or when we have a
few spare moments to help curious or confused users.  In
this monthly feature, we highlight some of the top-voted questions and
answers posted since our last update.*

{% comment %}<!-- https://bitcoin.stackexchange.com/search?tab=votes&q=created%3a1m..%20is%3aanswer -->{% endcomment %}
{% assign bse = "https://bitcoin.stackexchange.com/a/" %}

FIXME:bitschmidty

## Releases and release candidates

*New releases and release candidates for popular Bitcoin infrastructure
projects.  Please consider upgrading to new releases or helping to test
release candidates.*

- [HWI 2.0.1][] is a bugfix release that addresses minor issues with Trezor
  T passphrase entry and keyboard shortcuts in the `hwi-qt` user
  interface.

- [C-Lightning 0.10.0-rc1][C-Lightning 0.10.0] is a release candidate
  for the next major version of this LN node software.

> 

Reviewer 2: Just released, if interesting:
- Specter desktop 1.3.0 https://github.com/cryptoadvance/specter-desktop/releases/tag/v1.3.0

Reviewer 2: this one might be important 

https://github.com/btcpayserver/btcpayserver/releases/tag/v1.0.7.1

"This is a security release that patches one critical and several low-impact vulnerabilities that affected BTCPay Server versions 1.0.7.0 and older."

Author: @jonatack indeed!  Thanks!

   rebalancing, they estimate a 48% improvement.  Given that each retry
   usually requires several seconds, and could take much longer in some
   cases, this could provide a significant benefit to the network.
   The algorithm was tested against several example networks, including
   one drawn from a snapshot of almost 1,000 live channels.

   The paper deliberately does not take routing fees into consideration,
   and most mailing list discussion about the topic focused on how to
   use the results while still ensuring users don't pay an excessive
   amount of fees.

## Selected Q&A from Bitcoin StackExchange

*[Bitcoin StackExchange][bitcoin.se] is one of the first places Optech
contributors look for answers to their questions---or when we have a
few spare moments to help curious or confused users.  In
this monthly feature, we highlight some of the top-voted questions and
answers posted since our last update.*

{% comment %}<!-- https://bitcoin.stackexchange.com/search?tab=votes&q=created%3a1m..%20is%3aanswer -->{% endcomment %}
{% assign bse = "https://bitcoin.stackexchange.com/a/" %}

- [How hard is it for an exchange to adopt native segwit?]({{bse}}103674)

>   instagibbs lists a few considerations for exchanges implementing native segwit

Reviewer 2: ```suggestion
  instagibbs lists a few considerations for exchanges implementing native segwit,
```

  including address generation, compatibility of signing infrastructure like HSMs,

>   ensuring spendability, and support and business considerations.

Reviewer 2: it's unclear (to me) if the "ensuring..." part is in the same list or a subordinate clause to "like HSMs"

"HSMs" doesn't seem to be an unintroduced abbreviation

Author: @bitschmidty regarding the "ensuring..." feedback, one of the writer tricks that mostly avoids this problem is to always put list items in order of least to most number of syllables when no better sort order suggests itself.  E.g.: "address generation, ensuring spendability, support and business considerations, and compatibility of signing infrastructure like Hardware Security Modules (HSMs)."

(One exception to this rule is to sort most to least number of syllables when you want your list to end memorably, suddenly, and sharply.)

Reviewer 3: @jonatack thanks for pointing this out and thanks @harding for getting me out of my bind! Took the suggestion

- [How do you calculate when 98% of Bitcoin will be mined?]({{bse}}103159)

>   While Murch provides an approximation of either 2030 or 2031 for when 98% of

Reviewer 2: maybe
```suggestion
  While Murch provides an estimate of 2030-2031 for when 98% of
```

  Sheet][reward schedule google sheet] with additional metrics.

- [How can I use Bitcoin Core with the anonymous network protocol I2P?]({{bse}}103402)
  With the merge of [Bitcoin Core #20685][news139 i2p], Bitcoin supports the I2P
  network. Michael Folkson summarizes [Jon Atack's original thread][jonatack
  twitter i2p] on how to configure Bitcoin Core to use I2P.

- [Will nodes with a larger-than-default mempool retransmit transactions that have been dropped from smaller mempools?]({{bse}}103104)
  Pieter Wuille notes that transaction rebroadcasting is currently a [wallet
  responsibility][se 103261], that perhaps nodes should also rebroadcast unconfirmed
  transactions, and points to [Bitcoin Core #21061][] as working toward that goal.

- [Should block height or MTP or a mixture of both be used in a soft fork activation mechanism?]({{bse}}103854)
  David A. Harding outlines the advantages and disadvantages of both Median
  Time Past (MTP) and block heights as timing mechanisms within Bitcoin. MTP
  roughly corresponds to clock time but can manipulated by miners to skip a
  signaling period. Block heights are not consistent but are also not

>   miner-gamable in the same way as MTP.

Reviewer 2: ```suggestion
  miner-gameable in the same way as MTP.
```

  doc] for building Bitcoin Core for Android and a [continuous integration
  job][android ci] to test the Android build system.

- [Rust-Lightning #849][] makes a channel's `cltv_expiry_delta`
  configurable and reduces the default value from 72 blocks to 36
  blocks.  This parameter sets the minimum number of blocks a node will
  have to settle a payment attempt with its upstream peer after learning
  from its downstream peer whether or not that payment succeeded; it must
  be long enough to settle the payment onchain if necessary but should
  be short enough that it's competitive with other nodes that are trying
  to minimize possible delays.  See also [Newsletter #40][news40
  cltv_expiry_delta] where LND reduced its value to 40 blocks.

- [C-Lightning #4427][] makes it possible to experiment with
  dual funded payment channels by using the configuration option
  `--experimental-dual-fund`.  Dual funding allows funds for the initial
  channel balance to be contributed by both the node initiating the
  channel and the node accepting the channel, which can be useful for
  merchants and other users who want to begin receiving payments as
  soon as the channel finishes opening.

> - [Eclair #1738][] corrects the penalizing of revoked HTLC transactions by
>   spending all outputs spendable by the revocation key instead of just the
>   first. This is to account for the possibility that [anchor outputs][topic
>   anchor outputs] enabled peers merge their
>   `SIGHASH_SINGLE|SIGHASH_ANYONECANPAY` HTLC transactions, resulting in an
>   aggregate transaction with multiple outputs to be penalized.

Reviewer 4: What about:

```suggestion
- [Eclair #1738][] updates the penalty mechanism for [anchor outputs][topic anchor outputs]
  revoked HTLC transactions. Anchor outputs introduces the possibility
  to merge multiple `SIGHASH_SINGLE|SIGHASH_ANYONECANPAY` HTLC transactions,
  resulting in an aggregate transaction with multiple outputs to be penalized at once.
  The PR ensures that all outputs that are spendable with the revocation key are claimed
  instead of claiming only one per transaction.
```

Thanks again for highlighting this PR, it could be useful for other implementers who may not have noticed this edge case.

Reviewer: Change looks good to me. Thanks @t-bast!

Author: Made a different edit that incorporates several of @t-bast's suggested changes (thanks!).  In past writing ([e.g.](https://bitcoinops.org/en/newsletters/2020/12/02/#proposed-intermediate-solution-for-ln-sighash-single-fee-theft)) I've tried not to refer to the sighash changes as anchor outputs, since it's independent of the changes made to use CPFP carve out---it just happened to have been made in the same protocol change as anchor outputs.  My edited version tries to make that distinction (and also doesn't backlink to this item from our anchor outputs topic for that reason) and also adds a link to a recent news item where we mentioned HTLC enforcement batching.

Thanks @dongcarl for the writeup and @t-bast for not only reviewing but also suggesting edits!

  While Murch provides an estimate of 2030-2031 for when 98% of
  all bitcoins will be mined, he also links to a [reward schedule Google
  Sheet][reward schedule google sheet] with additional metrics.

- [How can I use Bitcoin Core with the anonymous network protocol I2P?]({{bse}}103402)
  With the merge of [Bitcoin Core #20685][news139 i2p], Bitcoin supports the I2P
  network. Michael Folkson summarizes [Jon Atack's original thread][jonatack
  twitter i2p] on how to configure Bitcoin Core to use I2P.

- [Will nodes with a larger-than-default mempool retransmit transactions that have been dropped from smaller mempools?]({{bse}}103104)
  Pieter Wuille notes that transaction rebroadcasting is currently a [wallet
  responsibility][se 103261], that perhaps nodes should also rebroadcast unconfirmed
  transactions, and points to [Bitcoin Core #21061][] as working toward that goal.

- [Should block height or MTP or a mixture of both be used in a soft fork activation mechanism?]({{bse}}103854)
  David A. Harding outlines the advantages and disadvantages of both Median
  Time Past (MTP) and block heights as timing mechanisms within Bitcoin. MTP

>   roughly corresponds to clock time but can manipulated by miners to skip a

Reviewer 5: ```suggestion
  roughly corresponds to clock time but can be manipulated by miners to skip a
```

---
title: 'Bitcoin Optech Newsletter #142'
permalink: /en/newsletters/2021/03/31/
name: 2021-03-31-newsletter
slug: 2021-03-31-newsletter
type: newsletter
layout: newsletter
lang: en
---
This week's newsletter describes a paper and a short discussion about
probabilistic path selection for LN and includes our regular sections with
summaries of popular questions and answers from the Bitcoin
StackExchange, release and release candidates, and notable changes to
Bitcoin infrastructure software.

## Action items

- **Upgrade BTCPay Server to 1.0.7.1:** this [release][btcpay server
  1.0.7.1] fixes "one critical and several low-impact vulnerabilities

>   that affected BTCPay Server versions 1.0.7.0 and older", according to

Reviewer: See previous discussion on "programmer quotes:" https://github.com/bitcoinops/bitcoinops.github.io/pull/506#discussion_r560536416
